### DISCUSSION TITLE
How to use pytest.mark.parametrize (or equivilant) inside a plugin

### DISCUSSION THREAD

**Discussion by Cameronsplaze** [pytest-dev/pytest#8893] [Q&A] [answered]
I have a conftest.py file, that's duplicated across repos, so I'm trying to turn it into a plugin. It parses yml files for tests, then adds those tests to the overall suite, using "@pytest.mark.parametrize". This part works in conftest.py!

When switching the code to the plugin, the parametrize function gets skipped completely. Is this expected? Is there a hook I can use to replicate the same behavior of adding tests?

I saw the pytest_generate_tests hook, but it's called for *each* test_.py file, (not just once). I could maybe get it to skip unless it hits my custom test_.py, but I'm not sure how to get it to call on a test_.py *inside* my plugin. That also seems like a hacky answer, so I'm hoping there's a better one?

Thanks in advance for any pointers on this!

**Comment by Cameronsplaze** [comment, marked as answer]
The best way I can find is to base it on this example here: https://docs.pytest.org/en/6.2.x/example/nonpython.html
There's a bit of the plugin I'm having to refactor to do so, but it's looking like it'll be cleaner in the end!

In the end, I couldn't find a way to use @pytest.mark.parametrize, but I'm not sure it's necessary here.

**Reply by nicoddemus** [reply]
`pytest.mark.parametrize` is meant to be used in tests only, how were you using it in the `conftest.py` file? Also moving from a `conftest.py` to a plugin should be seamless.

**Reply by Cameronsplaze** [reply]
The repo's here if you wanna see source: https://github.com/asfadmin/Discovery-PytestAutomation. The prod branch has the old code I'm refactoring, and the main branch is the new wip version.

I was grabbing a list of tests in a yml, and passing each one (with parametrize) to a method that figured out which test it was for, and passing the arguments onto that test. The core code, I don't have to re-write, but how it hooks into pytest I think I do.

I'm definitely already seeing some nice bonuses I'm getting for free, like the file names listing as the test run, and having repr_failure in the class exposed for uniform error reporting. I'm excited xD

-----
Edit: Looking at the old code closer, you're right: parametrize not in conftest.py. It's in test_MainManager.py, and I have a bunch of helper methods in conftest.py.
